[1301] Add support for inter model references

lfasani opened the pull request:
ticket #1300 and #1301

lfasani committed:
[1300] Rename SiriusWebJSONResourceFactoryImpl to JSONResourceFactory

Bug: https://github.com/eclipse-sirius/sirius-components/issues/1300
Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

lfasani committed:
[releng] Switch to EMFJson 2.3.2-SNAPSHOT

Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

lfasani committed:
[1301] Add support for intermodel references

The referenced object concrete type is now serialized in JSON resource
JSON resources should now be created with the right scheme

Bug: https://github.com/eclipse-sirius/sirius-components/issues/1301
Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

lfasani committed:
[1301] Add a cross reference adapter to clean the inverse references

This adapter specializes ECrossReferenceAdapter in case of deletion of
an EObject or a Resource in order to:
* clean the proxies of the inverse references
* clean the ECrossReferenceAdapter itself

Bug: https://github.com/eclipse-sirius/sirius-components/issues/1301
Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

pcdavid commented on CHANGELOG.adoc:
```diff
@@ -1,6 +1,20 @@
+
+=== Bug fixes
+
+- https://github.com/eclipse-sirius/sirius-components/issues/1301[#1301] [core] Add support for having intermodel references
```
Maybe mention "(inside the same project)".

lfasani replied:
done

pcdavid commented on CHANGELOG.adoc:
```diff
@@ -40,6 +54,7 @@ The `label` value is displayed before the widget to be consistent with other pro
 - [frontend] Switch to `@apollo/client 3.6.9`, `@material-ui/core 4.12.4`, `@material-ui/icons 4.11.3`, `graphql 16.5.0`, `prop-types 15.8.1`, `sprotty 0.12.2` and `xstate 4.32.1`. Various development dependencies have also been updated such as typescript, rollup, prettier etc
 - [backend] Switch to https://github.com/spring-projects/spring-boot/releases/tag/v2.7.2[Spring Boot 2.7.2]
 - [backend] Switch to the managed version of GraphQL Java, as of Spring Boot 2.7.2, we will thus use GraphQL Java 18.2
+- [backend] Switch to EMFJson 2.3.2-SNAPSHOT
```
This should be in the v2022.11.0 section.

lfasani replied:
done

pcdavid commented on CHANGELOG.adoc:
```diff
@@ -117,7 +132,7 @@ To integrate the "Details" view inside a workbench, one must now use the new `De
 - https://github.com/eclipse-sirius/sirius-components/issues/1141[#1141] [workbench] It is now possible to add a "Related View" contribution to the workbench.
 It behaves like the "Details" view in that it reacts to the selection by displaying a Form, but is designed to display contextual information about the selected element (i.e. which elements it points to and which elements point to it for example).
-The actual definition of the application (e.g. Sirius Web), which must provide a single bean implementing the new `IRelatedElementsDescriptionProvider` interface.
+The actual definition of the application (e.g. Sirius Web), which must provide a single bean implementing the new `IRelatedElementsDescriptionProvider` interface
```
Typo? No reason to remove the trailing dot.

lfasani replied:
done

pcdavid commented on doc/adrs/069_add_support_for_intermodel_reference.adoc:
```diff
@@ -0,0 +1,41 @@
+That is an issue because IllegalArgumentException are raised from some EMF code part (such as URI.resolve(URI)).
+A valid URI must contain a scheme.
+So the URI of the resources added to the resourceSet should be on the form:
+`<scheme>:///<id de la resource>`
```
"id de la resource" in french? :smile:

lfasani replied:
done

pcdavid commented on doc/adrs/069_add_support_for_intermodel_reference.adoc:
```diff
@@ -0,0 +1,41 @@
+
+=== The unresolvable proxies
+Let's say the EditingContext contains two resources A and B with aObj in A referencing bObj in B.
+Currently, if bObj or B is removed, aObj holds a proxy to bObj that can not be resolved.
```
Nitpicking, but technicaly just after bObj or B is removed aObj still references the exact same bObj instance as before (which is a real issue in itself, but does not involves proxies at this time). It's only when A will be serialized that it will store a reference to an object that will not be present anymore on the next reload and result in an unresolvable proxy.

lfasani replied:
I rewrote the text to be correct.

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java:
```diff
@@ -0,0 +1,121 @@
+    private void handleRemoveObject(Object object) {
+        if (object instanceof EObject) {
+            this.handleRemoveEObject((EObject) object);
+        } else if (object instanceof Resource) {
```
Maybe test `instanceof Resource` first? Just in case...

lfasani replied:
You played too much with CDO ;)
Done

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java:
```diff
@@ -0,0 +1,121 @@
+        Object oldValue = notification.getOldValue();
+        int eventType = notification.getEventType();
+
+        if (eventType == Notification.REMOVE || eventType == Notification.REMOVE_MANY) {
```
The enclosing `if (eventType == Notification.REMOVE || eventType == Notification.REMOVE_MANY)` seems unneeded.

lfasani replied:
Why? I do not want to do unnecessary work

pcdavid replied:
This would be simpler and not perform any test twice:

```java
if (eventType == Notification.REMOVE) {
  this.handleRemoveObject(oldValue);
} else if (eventType == Notification.REMOVE_MANY) {
  for (Object object : (Collection<?>) oldValue) {
    this.handleRemoveObject(object);
  }
} else {
  super.handleContainment(notification);
}
```

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java:
```diff
@@ -0,0 +1,121 @@
+        this.unsetTarget(resource);
+    }
+
+    private void removeProxyFromInverseReference(EObject referencedObject) {
```
Suggested name: `clearReferencesTo()`

lfasani replied:
done

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java:
```diff
@@ -0,0 +1,121 @@
+        Collection<Setting> settings = this.getInverseReferences(referencedObject);
+
+        if (settings != null) {
+            Collection<Setting> nonContainmentReferences = this.getNonContainmentReferences(settings);
```
This method (`removeProxyFromInverseReference`) will be called *a lot*. It should be as efficient as possible. It should do the iteration and test on the references itself instead of calling `getNonContainmentReferences` which allocates a new `ArrayList` every time.
Maybe it should even avoid calling `getInverseReferences` (which does quite a lot of work, some which may not be needed in our case) and use `ECrossReferenceAdapter.inverseCrossReferencer` directly.

lfasani replied:
I tried to `ECrossReferenceAdapter.inverseCrossReferencer` directly but it fails for the `EOpposite` features.
It is actually, the aim of `getNonContainmentReferences ` to manage the EOpposite

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java:
```diff
@@ -0,0 +1,121 @@
+ *
+ * @author lfasani
+ */
+public class EditingContextCrossReferenceAdapter extends ECrossReferenceAdapter {
```
I don't have a better option, but the performance impacts could be significant.

pcdavid replied:
We'll deal with performance issues *if* and when they actually occur.

pcdavid reviewed (changes requested):
Not actually tested yet in practice, but I publish my remarks anyway. More may come after testing.

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/JSONResourceFactory.java:
```diff
@@ -21,18 +21,43 @@
 
     @Override
     public JsonResource createResource(URI uri) {
         Map<String, Object> options = new HashMap<>();
```
I propose we add this here:
```java
if (uri == null || uri.scheme() == null || uri.scheme().isBlank()) {
            throw new IllegalArgumentException(String.format("Missing scheme for URI %s", uri)); //$NON-NLS-1$
        }
```
to make any misuse fail early. Even if we review and fix all currently existing callers in client code (Sirius Web and others), it's pretty easy for this kind of misuse to reappear later on.

lfasani replied:
done

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/JSONResourceFactory.java:
```diff
@@ -21,18 +21,43 @@
+     * @param resourceId
+     *            the id of the resource
+     */
+    public JsonResource createResource(String resourceId) {
```
The naming is confusing. It's the same name (`createResource`) for the version which takes a URI and the version which takes a "raw" resourceId.
It's pretty easy to use the wrong one. For example in `DeleteDocumentEventHandlerTests` we do:

```java
new JSONResourceFactory().createResource(URI.createURI(document.getId().toString()));
```

which creates a Resource with the wrong URI (`uri='a171a541-44cc-4607-9583-44fb6524e481'`). It should be:

```java
new JSONResourceFactory().createResource(document.getId().toString());
```

which looks very similar but behave very differently.

I wonder if `createResource(URI uri)` should just throw an `IllegalArgumentException` with a proper message if given a URI without a scheme. It would make most latent bugs obvious and identify the problematic caller immediatly.

lfasani replied:
I renamed `createResource(String)` to `createResourceFromPath`

pcdavid commented on CHANGELOG.adoc:
```diff
@@ -6,14 +6,19 @@
 - [ADR-070] How to contribute a new diagram node style
-- [ADR-071] - Add parametric SVG node style
+- [ADR-071] Add parametric SVG node style
+- [ADR-072] Add support for intermodel references(inside the same project)
```
Typo: missing space before "("

pcdavid commented on doc/adrs/072_add_support_for_intermodel_reference.adoc:
```diff
@@ -0,0 +1,41 @@
+= ADR-072 - Add support for intermodel references(inside a same project)
```
Same typo: missing space

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/JSONResourceFactory.java:
```diff
@@ -21,18 +21,47 @@
     @Override
     public JsonResource createResource(URI uri) {
+        if (uri == null || uri.scheme() == null || uri.scheme().isBlank()) {
+            throw new IllegalArgumentException(String.format("Missing scheme for URI %s", uri)); //$NON-NLS-1$
```
The explicit `throw` makes ArchUnit tests fail; should be rewritten to hide the throw (e.g. using an `Objects.requireNonNull` or `Optional#orElseThrow`...).

lfasani committed:
[1301] Adapt sirius-web code to support inter-model references

* Use a correct URI to instantiate a JSONResource. The URI should have a
scheme "sirius"
* Only resources with "sirius" scheme are persisted and displayed in the
Explorer
* This commit also ensures that, if a specific resource factory has been
registered, it will be used to instantiate the XMI resource at download
* Forbid the upload of a single document that contains proxies. The
check is not done when uploading a project.

https://github.com/eclipse-sirius/sirius-components/issues/1301
Signed-off-by: Laurent Fasani <laurent.fasani@obeo.fr>

pcdavid commented on packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/JSONResourceFactory.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public JsonResource createResource(URI uri) {
+        if (uri == null || uri.scheme() == null || uri.scheme().isBlank()) {
+            Optional.empty().orElseThrow(() -> new IllegalArgumentException(String.format("Missing scheme for URI %s", uri))); //$NON-NLS-1$
```
Maybe:
```java
Optional.ofNullable(uri).map(URI::scheme).orElseThrow(() -> ...)
```

lfasani replied:
done

pcdavid commented on packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/explorer/RenameDocumentTreeItemHandler.java:
```diff
@@ -71,7 +72,7 @@ public IStatus handle(IEditingContext editingContext, TreeItem treeItem, String
             // @formatter:off
             resourceSet.getResources().stream()
-                    .filter(resource -> documentEntity.getId().toString().equals(resource.getURI().toString()))
+                    .filter(resource -> new JSONResourceFactory().createResourceURI(documentEntity.getId().toString()).equals(resource.getURI()))
```
Seem wrong? Missing `.path().substring(1)`?

lfasani replied:
done

pcdavid commented on packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/documents/UploadDocumentEventHandler.java:
```diff
@@ -182,6 +191,41 @@ private String getContent(EPackage.Registry registry, UploadFile file) {
+        // @formatter:off
+        boolean resourceContainsAProxy = StreamSupport.stream(iterable.spliterator(), false)
+            .filter(eObject -> {
+                boolean eObjectcontainsAProxy = eObject.eClass().getEAllReferences().stream()
+                        .filter(ref -> !ref.isContainment() && eObject.eIsSet(ref))
+                        .filter(ref -> {
+                            boolean containsAProxy = false;
+                            Object value = eObject.eGet(ref);
+                            if (ref.isMany()) {
+                                List<?> list = (List<?>) value;
+                                containsAProxy = list.stream()
+                                        .filter(EObject.class::isInstance)
+                                        .map(EObject.class::cast)
+                                        .filter(EObject::eIsProxy)
+                                        .findFirst()
+                                        .isPresent();
+                            } else {
+                                containsAProxy = ((EObject) value).eIsProxy();
+                            }
+                            return containsAProxy;
+                        })
+                        .findFirst()
+                        .isPresent();
+
+                return eObjectcontainsAProxy;
```
Screams for a separate helper method, and just:
```java
.filter(this::hasProxyReferences)
```
